File: app/src/main/java/com/hippo/app/PrettyPreferenceActivity.java
/*
 * Copyright 2018 Hippo Seven
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hippo.app;

import android.content.Context;
import android.preference.PreferenceActivity;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.ListAdapter;
import android.widget.TextView;
import androidx.annotation.NonNull;
import com.hippo.ehviewer.R;
import java.util.ArrayList;
import java.util.List;

public abstract class PrettyPreferenceActivity extends AppCompatPreferenceActivity {

  @Override
  public void setListAdapter(ListAdapter adapter) {
    if (adapter == null) {
      super.setListAdapter(null);
      return;
    }

    int count = adapter.getCount();
    List<PreferenceActivity.Header> headers = new ArrayList<>(count);
    for (int i = 0; i < count; ++i) {
      headers.add((PreferenceActivity.Header) adapter.getItem(i));
    }

    super.setListAdapter(new HeaderAdapter(this, headers, R.layout.item_preference_header, true));
  }

  private static class HeaderAdapter extends ArrayAdapter<PreferenceActivity.Header> {
    private static class HeaderViewHolder {
      ImageView icon;
      TextView title;
      TextView summary;
    }

    private LayoutInflater mInflater;
    private int mLayoutResId;
    private boolean mRemoveIconIfEmpty;

    private HeaderAdapter(Context context, List<PreferenceActivity.Header> objects, int layoutResId,
        boolean removeIconBehavior) {
      super(context, 0, objects);
      mInflater = (LayoutInflater)context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
      mLayoutResId = layoutResId;
      mRemoveIconIfEmpty = removeIconBehavior;
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
      HeaderViewHolder holder;
      View view;

      if (convertView == null) {
        view = mInflater.inflate(mLayoutResId, parent, false);
        holder = new HeaderViewHolder();
        holder.icon = view.findViewById(android.R.id.icon);
        holder.title = view.findViewById(android.R.id.title);
        holder.summary = view.findViewById(android.R.id.summary);
        view.setTag(holder);
      } else {
        view = convertView;
        holder = (HeaderViewHolder) view.getTag();
      }

      // All view fields must be updated every time, because the view may be recycled
      PreferenceActivity.Header header = getItem(position);
      if (mRemoveIconIfEmpty) {
        if (header.iconRes == 0) {
          holder.icon.setVisibility(View.GONE);
        } else {
          holder.icon.setVisibility(View.VISIBLE);
          holder.icon.setImageResource(header.iconRes);
        }
      } else {
        holder.icon.setImageResource(header.iconRes);
      }
      holder.title.setText(header.getTitle(getContext().getResources()));
      CharSequence summary = header.getSummary(getContext().getResources());
      if (!TextUtils.isEmpty(summary)) {
        holder.summary.setVisibility(View.VISIBLE);
        holder.summary.setText(summary);
      } else {
        holder.summary.setVisibility(View.GONE);
      }

      return view;
    }
  }
}

File: app/src/main/java/com/hippo/app/PrettyPreferenceActivity.kt
/*
 * Copyright 2018 Hippo Seven
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hippo.app

import android.content.Context
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.ListAdapter
import android.widget.TextView
import com.hippo.ehviewer.R

abstract class PrettyPreferenceActivity : AppCompatPreferenceActivity() {
    @Deprecated("Deprecated in Java")
    override fun setListAdapter(adapter: ListAdapter) {
        if (adapter == null) {
            super.setListAdapter(null)
            return
        }
        val count = adapter.count
        val headers: MutableList<Header> = ArrayList(count)
        for (i in 0 until count) {
            headers.add(adapter.getItem(i) as Header)
        }
        super.setListAdapter(HeaderAdapter(this, headers, R.layout.item_preference_header, true))
    }

    private class HeaderAdapter constructor(
        context: Context, objects: List<Header>, layoutResId: Int,
        removeIconBehavior: Boolean
    ) : ArrayAdapter<Header?>(context, 0, objects) {
        private class HeaderViewHolder {
            var icon: ImageView? = null
            var title: TextView? = null
            var summary: TextView? = null
        }

        private val mInflater: LayoutInflater
        private val mLayoutResId: Int
        private val mRemoveIconIfEmpty: Boolean

        init {
            mInflater = context.getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
            mLayoutResId = layoutResId
            mRemoveIconIfEmpty = removeIconBehavior
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val holder: HeaderViewHolder
            val view: View
            if (convertView == null) {
                view = mInflater.inflate(mLayoutResId, parent, false)
                holder = HeaderViewHolder()
                holder.icon = view.findViewById(android.R.id.icon)
                holder.title = view.findViewById(android.R.id.title)
                holder.summary = view.findViewById(android.R.id.summary)
                view.tag = holder
            } else {
                view = convertView
                holder = view.tag as HeaderViewHolder
            }

            // All view fields must be updated every time, because the view may be recycled
            val header = getItem(position)
            if (mRemoveIconIfEmpty) {
                if (header!!.iconRes == 0) {
                    holder.icon!!.visibility = View.GONE
                } else {
                    holder.icon!!.visibility = View.VISIBLE
                    holder.icon!!.setImageResource(header.iconRes)
                }
            } else {
                holder.icon!!.setImageResource(header!!.iconRes)
            }
            holder.title!!.text = header.getTitle(context.resources)
            val summary = header.getSummary(context.resources)
            if (!TextUtils.isEmpty(summary)) {
                holder.summary!!.visibility = View.VISIBLE
                holder.summary!!.text = summary
            } else {
                holder.summary!!.visibility = View.GONE
            }
            return view
        }
    }
}